torch/_dynamo/output_graph.py

```diff
@@ -1183,6 +1206,131 @@ def remove_unused(node):
                     # Make sure we delete later occurrences of the same symbol
                     used_symbols.remove(symbol)
 
+    # TODO: this is a generic pass that should live outside of Dynamo
+    def insert_deferred_runtime_asserts(self, root, name) -> None:
+        """
+        During tracing, we may have discovered that some data-dependent values
+        had runtime assert on them; e.g., torch.empty(x.item()) induces a runtime
+        that x.item() >= 0.  This asserts can happen unpredictably during fake
+        tensor propagation, so we cannot conveniently insert them into the FX graph
+        when they occur.  Instead, we accumulate them in the ShapeEnv, and in this
+        pass insert them into the graph as proper tests.
```

Comment:
Why would we not structure these as guards?

It follows the same patterns of needing symbols from shape_env, and being deferred till the end. The benefit though is a lot less structural code here, and a lot easier to extend to other things?

On top of that, you could imagine non export

Reply:
They can't be guards because you cannot run them upfront: they depend on unbacked symints which don't come into existence until midway through runtime execution. This applies for both export as well as eager.

Reply:
Yeah, maybe I was not clear.

The *purpose* of these as runtime asserts midway through the graph is crystal clear - the word "structure" in my question is both load bearing and confusing, I apologize. More like, why cant we use existing guard structures and notions here? something like

guards, asserts = _produce_guards_and_asserts()

where we used to just have _produce_guards()

What prompted this questions was seeing `bound_symbols` tracking change in output_graph, it just felt somewhat like what we do in tracked_fakes already? feel free to ignore if you just prefer this separate structure.

Reply:
The `bound_symbols` tracking is an unrelated refactor that I loaded into the PR to avoid O(N^2) bookkeeping in detecting which placeholders map to particular symbolic ints.

We cannot use existing guard structures because they generate Python code. However, for things I place in the FX graph, I cannot insert Python code directly: I must insert FX nodes. (Well, I *could* eval code on the fly for insertion into the FX graph, but I like this way better!)

Reply:
yeah agreed